File: src/hangman.hs
import Control.Exception (try, SomeException)
import qualified Data.Set as Set
import System.Environment
import System.Exit
import System.IO
import System.Random

main :: IO ()
main = do
    putStrLn "=== HANGMAN ==="
    filename <- tryGetFilenameArg
    allWords <- tryReadFileLines filename
    putStrLn "\n\n\n\n\n"
    infiniteLoop allWords

tryGetFilenameArg :: IO String
tryGetFilenameArg = do
    args <- getArgs
    if length args < 1 then do
        hPutStrLn stderr "Please provide a file path."
        exitWith $ ExitFailure 1
    else do
        return $ head args

tryReadFileLines :: String -> IO [String]
tryReadFileLines filename = do
    result <- try $ readFile filename :: IO (Either SomeException String)
    case result of
        Left _ -> do
            hPutStrLn stderr "Failed to read file."
            exitWith $ ExitFailure 2
        Right file -> do
            return $ lines file

infiniteLoop :: [String] -> IO ()
infiniteLoop allWords = do
    index <- randomRIO (0, length allWords - 1)
    let word = allWords !! index
    let correct = Set.empty
    let incorrect = Set.empty

    gameLoop word correct incorrect
    infiniteLoop allWords

gameLoop :: String -> CharSet -> CharSet -> IO ()
gameLoop word correct incorrect = do
    putStr $ concat $ replicate 5 "\x1b[A\x1b[K"

    let visible = mapVisible correct word
    let isWin = not $ any (== '_') visible

    if isWin then do
        endScreen word "You win! :)"
    else if length incorrect >= 6 then do
        endScreen word "You lose! :("
    else do
        putStrLn visible
        putStrLn $ "Chances: " ++ (show $ 6 - length incorrect)
        putStrLn $ "Correct: " ++ (joinArray $ Set.toList correct)
        putStrLn $ "Incorrect: " ++ (joinArray $ Set.toList incorrect)
        putStr "Guess: "
        hFlush stdout

        line <- getLine
        let guess = head line

        if length line < 1 then do
            gameLoop word correct incorrect
        else if guess == '\n' then do
            return ()
        else if elem guess word then
            gameLoop word (Set.insert guess correct) incorrect
        else
            gameLoop word correct (Set.insert guess incorrect)

endScreen :: String -> String -> IO ()
endScreen word msg = do
    putStrLn "---------"
    putStrLn msg
    putStrLn $ "The word was: '" ++ word ++ "'"
    putStrLn "---------"
    _ <- getLine
    return ()

type CharSet = Set.Set Char

mapVisible :: CharSet -> String -> String
mapVisible correct = map (\ch -> if Set.member ch correct then ch else '_')

joinArray :: String -> String
joinArray []     = []
joinArray [x]    = [x]
joinArray (x:xs) = (x : ", " ++ joinArray xs)

File: src/hangman.lua
local function main()
    print("=== HANGMAN ===")

    if #arg < 1 then
        print("Please provide a file path.")
        os.exit(1)
    end
    local filename = arg[1]

    local words = ReadFileLines(filename)
    if words == nil then
        print("Failed to read file.")
        os.exit(2)
    end

    print("\n\n\n\n\n")

    while true do
        local index = math.random(1, #words)
        local word = words[index]

        local correct = ""
        local incorrect = ""

        while true do
            for _ = 1, 5 do
                io.write("\033[A\033[K");
            end

            local visible = ""
            local is_win = true
            for i = 1, #word do
                local ch = word:sub(i, i)
                if string.find(correct, ch) then
                    visible = visible .. ch
                else
                    visible = visible .. "_"
                    is_win = false
                end
            end

            if is_win then
                print("---------")
                print("You win! :)")
                print("The word was: '" .. word .. "'")
                print("---------")
                _ = io.read()
                break
            end
            if #incorrect >= 6 then
                print("---------")
                print("You lose! :(")
                print("The word was: '" .. word .. "'")
                print("---------")
                _ = io.read()
                break
            end

            print(visible)
            print("Chances: " .. (6 - #incorrect))
            print("Correct: " .. StringSeparate(correct))
            print("Incorrect: " .. StringSeparate(incorrect))
            io.write("Guess: ")

            local line = io.read()
            if #line >= 1 then
                local guess = line:sub(1, 1)

                if string.find(word, guess) then
                    if not string.find(correct, guess) then
                        correct = correct .. guess
                    end
                else
                    if not string.find(incorrect, guess) then
                        incorrect = incorrect .. guess
                    end
                end
            end
        end
    end
end

function ReadFileLines(filename)
    local file = io.open(filename, "r")
    if file == nil then
        return
    end

    local lines = {}
    for line in file:lines() do
        if line:sub(-1) == "\r" then
            line = line:sub(1, -2)
        end
        table.insert(lines, line)
    end
    file:close()

    return lines
end

function StringSeparate(input)
    local output = ""
    for i = 1, #input do
        if i > 1 then
            output = output .. ", "
        end
        output = output .. input:sub(i, i)
    end
    return output
end

main()

